open-sphere-plugins/ogc-server/src/main/java/io/opensphere/server/control/AbstractServerSourceController.java

```diff
@@ -331,9 +349,69 @@ protected void updateSource(IDataSource source)
         finally
         {
             myConfigLock.writeLock().unlock();
+            if (myReloadListenerMap.containsKey(source))
+            {
+                EventListener<AbstractSingleStateEvent> listener = myReloadListenerMap.get(source);
+                myReloadListenerMap.remove(source);
+                listener.notify(new AbstractSingleStateEvent()
+                {
+                    @Override
+                    public String getDescription()
+                    {
+                        return "Deactivation of " + source.getName() + " is done, finishing reload.";
+                    }
+                });
+            }
         }
     }
 
+    /** Reload sources which are marked as active, but are not actually activated. */
+    protected void reloadBadSources()
+    {
+        Set<String> activeSourceURLs = getActiveSourceURLs();
+        getSourceList().stream().forEach(source ->
+        {
+            if (source instanceof URLSource)
+            {
+                if (source.isActive() && !activeSourceURLs.contains(((URLSource)source).getURLString()))
+                {
+                    myReloadListenerMap.put(source, new EventListener<AbstractSingleStateEvent>()
+                    {
+                        @Override
+                        public void notify(AbstractSingleStateEvent event)
+                        {
+                            activateSource(source);
+                        }
+                    });
+                    deactivateSource(source);
+                }
+            }
+        });
+    }
+
+    /**
+     * Gets the set of active sources.
+     *
+     * @return the set of active sources
+     */
+    private Set<String> getActiveSourceURLs()
+    {
+        Set<String> activeURLs = new HashSet<String>();
+        DataGroupController dgc = MantleToolboxUtils.getMantleToolbox(myToolbox).getDataGroupController();
+        dgc.getActiveMembers(false).stream().forEach(dti ->
```

Comment:
I think you want to pass in true.

Reply:
In this case - I don't believe that's necessary. All urls are found without looking recursively as far as I can tell - looking at the members recursively will just give repeats of the same urls from what I have seen.

Reply:
If ArcGIS server wont refresh if only World_Imagery is active then I am ok with that.

Reply:
Right now if a server has no active layers it will refresh, if any layers are turned on then it won't refresh. Will that catch all the cases we're expecting?

Reply:
Yeah as long as ArcGIS server with only World_Imagery active will not refresh then we are good.

Reply:
That is how it behaves - great.